src/plugins/discover/public/application/angular/discover.js

```diff
@@ -238,7 +238,8 @@ function discoverController($route, $scope, Promise) {
 
     if (!_.isEqual(newStatePartial, oldStatePartial)) {
       $scope.$evalAsync(async () => {
-        if (oldStatePartial.index !== newStatePartial.index) {
+        // NOTE: this is also called when navigating from discover app to context app
+        if (newStatePartial.index && oldStatePartial.index !== newStatePartial.index) {
```

Comment:
This is an actual fix for Discover 
When navigating to context we were setting `isChangingIndexPattern = true` which skips clearing a session: 
https://github.com/elastic/kibana/blob/386afdca8ffc8b5c61aa0c2ce0ea1a3476fd0aa7/src/plugins/discover/public/application/angular/discover.js#L343

Won't happen anymore